lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3541,19 +3548,16 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     //   }
     // }
     //
-    // Note: we relax the restriction if the type we're casting to is a
-    // non-final class because it's possible that we might have a subclass
-    // that conforms to the protocol.
-    if (fromExistential && !toExistential) {
-      if (auto NTD = toType->getAnyNominal()) {
-        if (!toType->is<ClassType>() || NTD->isFinal()) {
-          auto protocolDecl =
-              dyn_cast_or_null<ProtocolDecl>(fromType->getAnyNominal());
-          if (protocolDecl &&
-              !conformsToProtocol(toType, protocolDecl, dc)) {
-            return failed();
-          }
-        }
+    // Except when the toType is a generic archetype, because then it may
+    // have protocol conformances we cannot know statically.
+    if (fromExistential && (!toExistential && !toArchetype)) {
```

Comment:
Great @hamishknight :) I just adjusted the archetype logic, but I'm not sure the existential check should be inside `couldDynamicallyConformToProtocol` that check to me is more specific to casting rules than checking conformance itself ... maybe this could be an early return there too but I'm really not sure about it. Can you elaborate on that `!e toExistential` part?

Reply:
@LucianoPAlmeida The existential check is similar to the archetype check, it's because the underlying concrete type may conform to protocols we don't know about statically. For example:

```
protocol P {}
protocol Q {}
struct S : P, Q {}

let p: P = S()
let q = p as? Q
```

We don't know a `Q` conforming type can necessarily conform to `P`, and IMO this therefore fits with the rest of the checks in `couldDynamicallyConformToProtocol`.

Reply:
@LucianoPAlmeida The existential check is similar to the archetype check, it's because the underlying concrete type may conform to protocols we don't know about statically. For example:

```
protocol P {}
protocol Q {}
struct S : P, Q {}

let p: P = S()
let q = p as? Q
```

We don't know whether a `Q` conforming type can necessarily conform to `P`, and IMO this therefore fits with the rest of the checks in `couldDynamicallyConformToProtocol`.

Reply:
Ah I see, that makes sense ... thank you @hamishknight :)

Reply:
Great, just made the changes @hamishknight

Reply:
Thank you! I have one more round of feedback, but it should hopefully be the last :)